### PR TITLE
add getpwnam,getpwuid,getgrnam,getgrgid

### PR DESCRIPTION
**Pull request by hopkiw**

### CODE REVIEW

File: src/nss/new_nss_oslogin.c [hunk 1]
```diff
@@ -241,3 +241,174 @@ int recvline(struct Buffer *const buffer) {
...
+static enum nss_status
+_nss_oslogin_getpwnam_r(const char *name, struct passwd *result, char *buffer,
+                        size_t buflen, int *errnop) {
+  int res;
+  struct Buffer mgr;
+  memset(&mgr, 0, sizeof(struct Buffer));
+
+  *errnop = 0;
+  mgr.bufsize = BUFSIZE;
+  mgr.buf = (char *)malloc(BUFSIZE);
```
Comment (winterfreshx) [inline src/nss/new_nss_oslogin.c:254]: Is this module only planned to be used only in short running processes? 

I am worried about memory leaks if this module is used in the future by Daemons or other long running tasks. There was a similar pattern in the previous CL. I just want to make sure we're not going to cause problems in the future with memory management if someone uses this module for other purposes.

File: src/nss/new_nss_oslogin.c [hunk 2]
```diff
@@ -241,3 +241,174 @@ int recvline(struct Buffer *const buffer) {
...
+  if ((recvline(&mgr)) < 0) {
+    free(mgr.buf);
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  if (mgr.buf[0] == '\n') {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  res = parsepasswd(mgr.buf,result,buffer,buflen);
```
Comment (winterfreshx) [inline src/nss/new_nss_oslogin.c:277]: Nit, style consistency to have space after ,

File: src/nss/new_nss_oslogin.c [hunk 3]
```diff
@@ -241,3 +241,174 @@ int recvline(struct Buffer *const buffer) {
...
+
+  res = parsegroup(mgr.buf,result,buffer,buflen);
+  if (res == 0) {
+    return NSS_STATUS_SUCCESS;
+  }
+  *errnop = res;
+  if (res == ERANGE) {
+    return NSS_STATUS_TRYAGAIN;
+  }
+  return NSS_STATUS_NOTFOUND;
```
Comment (winterfreshx) [inline src/nss/new_nss_oslogin.c:413]: This code is repeated except for the string that is sent. Please create a common helper method that takes in the string for sprintf and performs the rest of the common work for easier refactoring at a later time.

**Review by winterfreshx — CHANGES_REQUESTED** [review]

File: src/nss/new_nss_oslogin.c (hunk 1 above)
Comment (hopkiw) [reply src/nss/new_nss_oslogin.c]: this comment confused me, bc this function returns and has nothing short or long about it. but i realized that you are talking about some missing calls to free(). updated PR accordingly

File: src/nss/new_nss_oslogin.c (hunk 3 above)
Comment (hopkiw) [reply src/nss/new_nss_oslogin.c]: i think we should not do this - c doesn't lend itself to overloading functions like that well. better to have the four repetitions.

File: src/nss/new_nss_oslogin.c [hunk 4]
```diff
@@ -235,9 +239,189 @@ int recvline(struct Buffer *const buffer) {
...
+static enum nss_status
+_nss_oslogin_getpwnam_r(const char *name, struct passwd *result, char *buffer,
+                        size_t buflen, int *errnop) {
+  int res;
+  struct Buffer mgr;
+  memset(&mgr, 0, sizeof(struct Buffer));
+
+  *errnop = 0;
+
+  if (dial(&mgr) != 0) {
```
Comment (Sarsate) [inline src/nss/new_nss_oslogin.c:260]: Do you want to set errno for failures in dial/send? As it is here, the behavior is identical to receiving an empty response. That might be what you want, but just want to make sure you've considered it. 

https://www.gnu.org/software/libc/manual/html_node/NSS-Modules-Interface.html

File: src/nss/new_nss_oslogin.c [hunk 5]
```diff
@@ -235,9 +239,189 @@ int recvline(struct Buffer *const buffer) {
...
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  // send the verb GETPWNAM with the argument <name>
+  // TODO: validate incoming length of 'name' fits in 100 char
+  char str[100];
```
Comment (Sarsate) [inline src/nss/new_nss_oslogin.c:266]: should this be a constant? why 100?

File: src/nss/new_nss_oslogin.c [hunk 6]
```diff
@@ -235,9 +239,189 @@ int recvline(struct Buffer *const buffer) {
...
+  memset(&mgr, 0, sizeof(struct Buffer));
+  *errnop = 0;
+
+  if (dial(&mgr) != 0) {
+    return NSS_STATUS_NOTFOUND;
+  }
+
+  // send the verb GETPWNAM with the argument <name>
+  // TODO: validate incoming length of 'name' fits in 100 char
+  char str[1000];
```
Comment (Sarsate) [inline src/nss/new_nss_oslogin.c:355]: 100?

**Review by Sarsate — APPROVED** [review]
LGTM

**Pull request merged**